packages/remix-dev/__tests__/routesConvention-test.ts

```diff
@@ -36,12 +36,68 @@ describe("createRoutePath", () => {
       ["[index]", "index"],
       ["test/inde[x]", "test/index"],
       ["[i]ndex/[[].[[]]", "index/[/[]"],
+
+      // Optional segment routes
+      ["(routes)/$", "routes?/*"],
+      ["(routes)/(sub)/$", "routes?/sub?/*"],
+      ["(routes).(sub)/$", "routes?/sub?/*"],
+      ["(routes)/($slug)", "routes?/:slug?"],
+      ["(routes)/sub/($slug)", "routes?/sub/:slug?"],
+      ["(routes).sub/($slug)", "routes?/sub/:slug?"],
+      ["(nested)/$", "nested?/*"],
+      ["(flat).$", "flat?/*"],
+      ["($slug)", ":slug?"],
+      ["(nested)/($slug)", "nested?/:slug?"],
+      ["(flat).($slug)", "flat?/:slug?"],
+      ["flat.(sub)", "flat/sub?"],
+      ["__layout/(test)", "test?"],
+      ["__layout.(test)", "test?"],
+      ["__layout/($slug)", ":slug?"],
+      ["(nested)/__layout/($slug)", "nested?/:slug?"],
+      ["($slug[.]json)", ":slug.json?"],
+      ["(sub)/([sitemap.xml])", "sub?/sitemap.xml?"],
+      ["(sub)/[(sitemap.xml)]", "sub?/(sitemap.xml)"],
+      ["(posts)/($slug)/([image.jpg])", "posts?/:slug?/image.jpg?"],
+      [
+        "($[$dollabills]).([.]lol)[/](what)/([$]).$",
+        ":$dollabills?/.lol)/(what?/$?/*",
```

Comment:
I left this one on purpose.

This one is a big edge case as mentioned above coming from allowing escaping of `[/]`:
https://github.com/remix-run/remix/pull/4706#discussion_r1036057053

I'm not counting it as a separator in those cases, so it is not considered an optional route. And adding code for this edge case would complicated the code.

I think it should be considered in another PR to throw an error if escaping separators shouldn't be allowed or would you suggest something else?

Reply:
Oh I see, `([.]lol)[/](what)` is _one_ path segment because the `/` is escaped. That leads to treating the outer `()` as a pair for denoting optionality (since we don't support a part of a segment being optional, its all or nothing).